### PR TITLE
SDK-2216: Added extended classes checker

### PR DESCRIPTION
**Pull request by sergeyspryker**
https://spryker.atlassian.net/browse/SDK-2216

### CODE REVIEW

File: src/Upgrade/Application/Strategy/ReleaseApp/Processor/EventSubscriber/CheckerExecutorEventSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+declare(strict_types=1);
+
+namespace Upgrade\Application\Strategy\ReleaseApp\Processor\EventSubscriber;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Upgrade\Application\Checker\ClassExtendsUpdatedPackageChecker\Synchronizer\PackagesSynchronizerInterface;
+use Upgrade\Application\Strategy\ReleaseApp\Processor\Event\ReleaseGroupProcessorEvent;
+
+class CheckerExecutorEventSubscriber implements EventSubscriberInterface
```
Comment (VladislavStrelchenko) [inline src/Upgrade/Application/Strategy/ReleaseApp/Processor/EventSubscriber/CheckerExecutorEventSubscriber.php:16]: What about `EvaluatorEventSubscriber` or `DynamicEvaluator...`?

File: src/Upgrade/Application/Checker/ClassExtendsUpdatedPackageChecker/ClassExtendsUpdatedPackageChecker.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        $violations = [];
+
+        foreach ($projectExtendedClasses as $projectExtendedClass => $fileName) {
+            if (!isset($vendorChangedClasses[$projectExtendedClass])) {
+                continue;
+            }
+
+            $fileName = strpos($fileName, $this->configurationProvider->getRootPath()) === 0 ? substr($fileName, $rootPathLength) : $fileName;
+
+            $violations[] = new ViolationDto('Project class extends class from updated package', $fileName, $vendorChangedClasses[$projectExtendedClass]);
```
Comment (VladislavStrelchenko) [inline src/Upgrade/Application/Checker/ClassExtendsUpdatedPackageChecker/ClassExtendsUpdatedPackageChecker.php:70]: Please move the message to Const

File: src/Upgrade/Application/Checker/ClassExtendsUpdatedPackageChecker/Fetcher/VendorChangedClassesFetcherInterface.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+declare(strict_types=1);
+
+namespace Upgrade\Application\Checker\ClassExtendsUpdatedPackageChecker\Fetcher;
+
+interface VendorChangedClassesFetcherInterface
+{
+    /**
+     * @return array<string, string> Key - className. Value - packageName
+     */
+    public function fetchVendorChangedClasses(): array;
```
Comment (VladislavStrelchenko) [inline src/Upgrade/Application/Checker/ClassExtendsUpdatedPackageChecker/Fetcher/VendorChangedClassesFetcherInterface.php:17]: Better rename to `fetchVendorChangedClassesWithPackage?`

File: src/Upgrade/Application/Checker/CheckerInterface.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+/**
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+declare(strict_types=1);
+
+namespace Upgrade\Application\Checker;
```
[the remark this answers is not in the document]
Comment (sergeyspryker) [reply src/Upgrade/Application/Checker/CheckerInterface.php:10]: What is for? It's part of upgrade process

File: src/Upgrade/Application/Strategy/ReleaseApp/Processor/EventSubscriber/CheckerExecutorEventSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+     * @param \Upgrade\Application\Strategy\ReleaseApp\Processor\Event\ReleaseGroupProcessorEvent $event
+     *
+     * @return void
+     */
+    public function onPostRequire(ReleaseGroupProcessorEvent $event): void
+    {
+        $stepsExecutorDto = $event->getStepsExecutionDto();
+
+        foreach ($this->checkers as $checker) {
+            $violations = $checker->check();
```
Comment (VladislavStrelchenko) [inline src/Upgrade/Application/Strategy/ReleaseApp/Processor/EventSubscriber/CheckerExecutorEventSubscriber.php:83]: We have to use enabler `ConfigurationProvider:isEvaluatorEnabled()`

**Review by VladislavStrelchenko — APPROVED** [review]

File: src/Core/Infrastructure/Service/Filesystem.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @throws \Symfony\Component\Filesystem\Exception\IOException
+     *
+     * @return array<string>
+     */
+    public function scanDir(string $directory, int $sorting_order = SCANDIR_SORT_ASCENDING, $context = null): array
+    {
+        $data = $context === null
+            ? static::box('scandir', $directory, $sorting_order)
+            : static::box('scandir', $directory, $sorting_order, $context);
```
Comment (DmytroKlymanSpryker) [inline src/Core/Infrastructure/Service/Filesystem.php:60]: You don't need to check here if context is null or not, it is null by default in scandir(). So it doesn't change anything if you will always call: 

```
$data = static::box('scandir', $directory, $sorting_order, $context);
```

The same is valid for file_get_contents()

**Review by DmytroKlymanSpryker — APPROVED** [review]

**Review by pavelmaksimov25 — APPROVED** [review]

**Pull request merged**